### PR TITLE
Compatibility mode

### PR DESCRIPTION
**Pull request by asenina**

### CODE REVIEW

File: nncf/quantization/algo.py [hunk 1]
```diff
@@ -359,14 +366,18 @@ def _quantize_activations(self, target_model: NNCFNetwork) -> List[InsertionComm
             original_nncf_graph = target_model.get_original_graph()
             for insertion_info, quantizer_config_list in insertion_data.items():
                 ia_op_exec_context = insertion_info.op_exec_context.input_agnostic
+                module_scope = str(ia_op_exec_context)
```
Comment (vshampor) [inline nncf/quantization/algo.py:369]: This variable does not correspond to a scope of a module, or even to a string representation of a module scope.

File: nncf/quantization/init_precision.py [hunk 2]
```diff
@@ -168,33 +169,33 @@ def apply_init(self):
...
+                hawq_debugger.dump_density_of_quantization_noise()
+                hawq_debugger.dump_perturbations_ratio()
+                hawq_debugger.dump_bitwidth_graph(self._algo, self._model)
+
+                self._model.rebuild_graph()
+                str_bw = [str(element) for element in self.get_bitwidth_per_scope()]
+                nncf_logger.info('\n'.join(['\n\"bitwidth_per_scope\": [', ',\n'.join(str_bw), ']']))
+
+                self._model.to(original_device)
+                return ordered_metric_per_layer
```
Comment (vshampor) [inline nncf/quantization/init_precision.py:198]: What is this whitespace change all about?

File: tests/helpers.py [hunk 3]
```diff
@@ -155,7 +155,7 @@ def create_compressed_model_and_algo_for_test(model: NNCFNetwork, config: NNCFCo
     -> Tuple[NNCFNetwork, CompressionAlgorithmController]:
     assert isinstance(config, NNCFConfig)
     NNCFConfig.validate(config)
-    algo, model = create_compressed_model(model, config, dump_graphs=False, dummy_forward_fn=dummy_forward_fn)
+    algo, model = create_compressed_model(model, config, dump_graphs=True, dummy_forward_fn=dummy_forward_fn)
```
Comment (vshampor) [inline tests/helpers.py:158]: Hope you change this back in the end.

File: tests/quantization/test_algo_quantization.py [hunk 4]
```diff
@@ -194,22 +197,6 @@ def forward(self, x):
...
-        "quantize_inputs": False})
-
-    _, algo = create_compressed_model_and_algo_for_test(model_cls(), config)
-
-    actual_pairs = algo.get_weights_activation_quantizers_pairs()
-    ref_pair_names = [(['NNCFConv2d[conv1]module_weight', 'NNCFConv2d[conv2]module_weight'],
-                       'ReLU[relu]/RELU_0')]
-
-    compare_weights_activation_quantizers_pairs(actual_pairs, algo, ref_pair_names, model_name)
-
```
Comment (vshampor) [inline tests/quantization/test_algo_quantization.py:212]: Why was this test removed?

File: tests/quantization/test_algo_quantization.py [hunk 5]
```diff
@@ -240,7 +227,7 @@ def test_can_load_quant_algo__with_defaults():
 
 def test_can_create_quant_loss_and_scheduler():
     config = get_quantization_config_without_range_init()
-    _, compression_ctrl = create_compressed_model_and_algo_for_test(MockModel(), config)
+    _, compression_ctrl = create_compressed_model_and_algo_for_test(BasicConvTestModel(), config)
```
Comment (vshampor) [inline tests/quantization/test_algo_quantization.py:230]: Why was the `MockModel` not enough for this test case?

File: tests/quantization/test_algo_quantization.py [hunk 6]
```diff
@@ -447,21 +434,13 @@ def test_quantize_inputs():
 
     model, compression_ctrl = create_compressed_model_and_algo_for_test(model, config)
     REF_QUANTIZED_INPUT_MODULE_SCOPES = [
```
Comment (vshampor) [inline tests/quantization/test_algo_quantization.py:436]: These are no longer "module scopes". Rename appropriately.

File: tests/quantization/test_onnx_export.py [hunk 7]
```diff
@@ -52,6 +52,7 @@ def test_onnx_export_to_fake_quantize(tmp_path):
 def test_onnx_export_to_quantize_dequantize(tmp_path):
     model = TwoConvTestModel()
     nncf_config = get_config_for_export_mode(should_be_onnx_standard=True)
+    nncf_config['quantizer_setup_type'] = 'pattern_based'
```
Comment (vshampor) [inline tests/quantization/test_onnx_export.py:55]: Should test the default method here; would be even better if we tested both.

File: tests/quantization/test_precision_init.py [hunk 8]
```diff
@@ -119,7 +119,8 @@ def create_hawq_test_config(batch_size=10, num_data_points=100):
 
 def create_hawq_hw_test_config(batch_size):
     config = create_hawq_test_config(batch_size)
-    config["hw_config_type"] = HWConfigType.VPU.value
+    config['quantizer_setup_type'] = "propagation_based"
```
Comment (vshampor) [inline tests/quantization/test_precision_init.py:122]: Don't we have propagation by default now?

File: tests/quantization/test_precision_init.py [hunk 9]
```diff
@@ -384,7 +387,7 @@ def test_hawq_broadcast_avg_traces_in_distributed_mode(tmp_path):
     num_data_points = 10
     batch_size = 2
     config = create_hawq_test_config(batch_size, num_data_points)
-
+    config.device = 'cuda'
```
Comment (vshampor) [inline tests/quantization/test_precision_init.py:390]: Why is this change required?

File: tests/quantization/test_precision_init.py [hunk 10]
```diff
@@ -418,16 +421,15 @@ def test_hawq_broadcast_avg_traces_in_distributed_mode(tmp_path):
 ]
 
 
-@pytest.mark.parametrize('hw_config', [None, HWConfigType.VPU],
-                         ids=['no_constraints', 'vpu_constraints'])
```
Comment (vshampor) [inline tests/quantization/test_precision_init.py:422]: Why was this parametrization removed?

File: tests/quantization/test_precision_init.py [hunk 11]
```diff
@@ -438,15 +440,34 @@ def test_hawq_manual_configs(manual_config_params, hw_config):
...
+
+    mocked_trace = mocker.patch('nncf.quantization.init_precision.HAWQPrecisionInitializer'
+                                '._filter_configs_by_precision_constraints')
+    mocked_trace.return_value = None
+
+    with pytest.warns(RuntimeWarning):
+        create_compressed_model_and_algo_for_test(model, config)
+
+
+def test_hawq_raises_error_if_method_returns_none(mocker):
```
Comment (vshampor) [inline tests/quantization/test_precision_init.py:460]: So what is the right behaviour that we test for? `_warns__if_method_returns_none` or `_raises_error_if_method_returns_none`? These are mutually exclusive definitions, aren't they?

File: tests/quantization/test_range_init.py [hunk 12]
```diff
@@ -455,6 +452,8 @@ def __len__(self):
             "input_info": {
                 "sample_size": [1, 3, 100, 100]
             },
+            "quantizer_setup_type": 'pattern_based',
+            "target_device": "NONE",
```
Comment (vshampor) [inline tests/quantization/test_range_init.py:456]: These two definitions should be mutually exclusive.

File: tests/quantization/test_range_init.py [hunk 13]
```diff
@@ -524,7 +523,9 @@ def test_range_init_is_called(config_cutter, range_init_call_count, precision_in
                               bn_adaptation_call_count, mocker):
     config = create_hawq_test_config()
     config['compression']['initializer'].update({'batchnorm_adaptation': {'num_bn_adaptation_steps': 5}})
-    model = MockModel()
+    config['input_info'] = {"sample_size": [1, 1, 4, 4]}
+    config.device = 'cuda'
+    model = BasicConvTestModel()
```
Comment (vshampor) [inline tests/quantization/test_range_init.py:528]: Same here - why was `MockModel` not enough?

File: tests/quantization/test_scheduler.py [hunk 14]
```diff
@@ -221,6 +221,7 @@ def test_staged_scheduler_with_hawq():
             }
         }
     })
+    config.device = 'cuda'
```
Comment (vshampor) [inline tests/quantization/test_scheduler.py:224]: Same

File: tests/quantization/test_unified_scales.py [hunk 15]
```diff
@@ -510,7 +511,8 @@ def test_unified_scales_for_vpu():
             "sample_size": [1, 1, 1, 1],
         }
     ]
-    nncf_config["hw_config_type"] = "vpu"
+    nncf_config["quantizer_setup__type"] = "propagation_based"
```
Comment (vshampor) [inline tests/quantization/test_unified_scales.py:514]: Should be default already

File: tests/quantization/test_unified_scales.py [hunk 16]
```diff
@@ -552,7 +554,8 @@ def test_unified_scales_are_identical_in_onnx(tmp_path):
             "sample_size": [1, 1, 1, 2],
         },
     ]
-    nncf_config["hw_config_type"] = "vpu"
+    nncf_config["quantizer_setup__type"] = "propagation_based"
```
Comment (vshampor) [inline tests/quantization/test_unified_scales.py:557]: Same

File: tests/test_sanity_sample.py [hunk 17]
```diff
@@ -389,6 +389,7 @@ def test_get_default_weight_decay(algo, ref_weight_decay):
 
 def test_cpu_only_mode_produces_cpu_only_model(config, tmp_path, mocker):
     config_factory = ConfigFactory(config['nncf_config'], tmp_path / 'config.json')
+    config['nncf_config']['device'] = 'cpu'
```
Comment (vshampor) [inline tests/test_sanity_sample.py:392]: Why was this change required?

**Review by vshampor — CHANGES_REQUESTED** [review]

File: nncf/quantization/init_precision.py (hunk 2 above)
Comment (ljaljushkin) [reply nncf/quantization/init_precision.py:198]: @vshampor, As far as I can see, Anastasia moved the whole block to the right under else condition. 
If all bit-width settings, generated by HAWQ, are incompatible with HW config, precision initialization will do nothing.

Definitely, this behavior should be fixed. By default, the error should be raised. But looks like it tests for manual HAWQ will fail. @asenina, am I right? If so, I can fix them later and this change is OK.

File: examples/classification/configs/quantization/densenet161_imagenet_custom_quant_pattern.json [hunk 18]
```diff
@@ -8,11 +8,12 @@
     "num_classes": 1000,
     "batch_size": 32,
     "epochs": 1,
-    "multiprocessing_distributed": true,
+    "multiprocessing_distributed": false,
     "optimizer": {
         "type": "Adam",
         "base_lr": 1e-6
     },
+    "quantizer_setup_type": "propagation_based",
```
[the remark this answers is not in the document]
Comment (asenina) [reply examples/classification/configs/quantization/densenet161_imagenet_custom_quant_pattern.json]: Done.

File: examples/classification/main.py [hunk 19]
```diff
@@ -125,7 +125,7 @@ def main_worker(current_gpu, config: SampleConfig):
     train_loader = train_sampler = val_loader = None
     resuming_checkpoint_path = config.resuming_checkpoint_path
     nncf_config = config.nncf_config
-
+    nncf_config.device = config.device
```
[the remark this answers is not in the document]
Comment (asenina) [reply examples/classification/main.py]: I changed the signature of the function `register_default_init_args` and now it is really unnecessary.

File: nncf/config_schema.py [hunk 20]
```diff
@@ -654,13 +656,7 @@ def with_attributes(schema: Dict, **kwargs) -> Dict:
...
     "required": ["input_info"],
-    "definitions": REF_VS_ALGO_SCHEMA,
-    "dependencies": {
-        "hw_config_type": {
-            "properties": {
-                "quantizer_setup_type": {"const": "propagation_based"}}
-        }
-    }
+    "definitions": REF_VS_ALGO_SCHEMA
 }
```
[the remark this answers is not in the document]
Comment (asenina) [reply nncf/config_schema.py:672]: We should set `quanizer_setup_type` = 'pattern_based' from config.  Please, take another look  at the changes in `nncf/model_creation.py`.

File: nncf/hw_config.py [hunk 21]
```diff
@@ -26,10 +26,19 @@
...
 
 
+
+HW_CONFIG_TYPE_TARGET_DEVICE_MAP = {
+    'ANY': 'CPU',
+    'CPU': 'CPU',
+    'VPU': 'VPU',
+    'GPU': 'GPU',
+    'NONE': None
+}
```
[the remark this answers is not in the document]
Comment (asenina) [reply nncf/hw_config.py]: Done.

File: nncf/initialization.py [hunk 22]
```diff
@@ -219,10 +219,15 @@ def _apply_initializers(self):
...
+                                                                      device=nncf_config.device),
+                                            BNAdaptationInitArgs(data_loader=train_loader,
+                                                                 device=nncf_config.device)])
     else:
-        nncf_config.register_extra_structs([QuantizationRangeInitArgs(data_loader=train_loader),
-                                            BNAdaptationInitArgs(data_loader=train_loader)])
+        nncf_config.register_extra_structs([QuantizationRangeInitArgs(data_loader=train_loader,
+                                                                      device=nncf_config.device),
+                                            BNAdaptationInitArgs(data_loader=train_loader,
+                                                                 device=nncf_config.device)])
```
[the remark this answers is not in the document]
Comment (asenina) [reply nncf/initialization.py]: Fixed.

File: nncf/quantization/algo.py [hunk 23]
```diff
@@ -233,7 +235,7 @@ def __get_scoped_quantizer_config(self, target_model: NNCFNetwork,
             else:
                 raise RuntimeError("Unable to use per channel quantization for module {} activations -"
                                    " input shape is unknown".format(parent_module_scope_str))
-        return qconfig
+        return module_scope_in_scope_overrides, qconfig
```
[the remark this answers is not in the document]
Comment (asenina) [reply nncf/quantization/algo.py]: Done

File: nncf/quantization/algo.py [hunk 24]
```diff
@@ -296,8 +298,13 @@ def _quantize_weights(self, target_model: NNCFNetwork) -> List[InsertionCommand]
...
             if self.hw_config is not None:
                 try:
+                    is_matched_on_scope_overrides, qconfig =\
+                     self.__get_scoped_quantizer_config(target_model, str(module_scope),
+                                                         is_weights=True)
+                    qconfig_overrides = qconfig if is_matched_on_scope_overrides else None
                     qconfig = self._select_final_qconfig(qconfig_list,
-                                                         self.global_quantizer_contraints[QuantizerGroup.WEIGHTS])
+                                                         self.global_quantizer_contraints[QuantizerGroup.WEIGHTS],
+                                                         qconfig_overrides)
```
[the remark this answers is not in the document]
Comment (asenina) [reply nncf/quantization/algo.py:305]: Done.

File: nncf/quantization/algo.py (hunk 1 above)
Comment (asenina) [reply nncf/quantization/algo.py]: Done.

File: tests/quantization/test_onnx_export.py (hunk 7 above)
Comment (asenina) [reply tests/quantization/test_onnx_export.py]: Done.

File: tests/quantization/test_precision_init.py (hunk 8 above)
Comment (asenina) [reply tests/quantization/test_precision_init.py]: Done.

File: tests/quantization/test_precision_init.py (hunk 9 above)
Comment (asenina) [reply tests/quantization/test_precision_init.py]: Fixed.

File: tests/quantization/test_precision_init.py (hunk 11 above)
Comment (asenina) [reply tests/quantization/test_precision_init.py]: Fixed.

File: tests/quantization/test_range_init.py (hunk 12 above)
Comment (asenina) [reply tests/quantization/test_range_init.py:453]: Fixed.

File: tests/quantization/test_scheduler.py (hunk 14 above)
Comment (asenina) [reply tests/quantization/test_scheduler.py]: Done.

File: tests/quantization/test_unified_scales.py (hunk 15 above)
Comment (asenina) [reply tests/quantization/test_unified_scales.py]: Done.

File: tests/quantization/test_unified_scales.py (hunk 16 above)
Comment (asenina) [reply tests/quantization/test_unified_scales.py]: Done.

File: tests/test_sanity_sample.py (hunk 17 above)
Comment (asenina) [reply tests/test_sanity_sample.py]: Fixed.

File: tests/quantization/test_algo_quantization.py (hunk 4 above)
Comment (ljaljushkin) [reply tests/quantization/test_algo_quantization.py:212]: no need to remove it, until pattern-based mode removal

File: tests/data/reference_graphs/quantized/symmetric/densenet121.dot [hunk 25]
```diff
@@ -1,1523 +1,1399 @@
 strict digraph  {
```
Comment (vshampor) [inline tests/data/reference_graphs/quantized/symmetric/densenet121.dot:1]: This graph is in a pretty bad shape - the concat'ed tensor WILL contain FP32 values and the batch norm result is not quantized at all. How we should handle this "separated" batch norm situation is an open question.

The situation could be made clearer by using pre-hooks. Skip this graph in pytest as well and leave the .dot files as-is.

Screenshot from 2020-10-13 15-10-26

Comment (vshampor) [reply tests/data/reference_graphs/quantized/symmetric/densenet121.dot:1]: Ok, found another issue here that had better be fixed immediately - the mergeable subgraphs get merged even though there is an output edge in the middle of the subgraph, see below. The merge should not have happened at all. I will fix this separately, wait for my fix.

Comment (vshampor) [reply tests/data/reference_graphs/quantized/symmetric/densenet121.dot:1]: Discussed with @alexsu52 - should probably make BatchNorm an INPUTS_QUANTIZABLE operation as well. The propagation mode will trigger a "wildcard" quantization for it since there is no "ScaleShift" or batch norm operation in the IR.

**Review by vshampor — CHANGES_REQUESTED** [review]

File: tests/data/reference_graphs/quantized/symmetric/densenet121.dot (hunk 25 above)
Comment (vshampor) [reply tests/data/reference_graphs/quantized/symmetric/densenet121.dot:1]: #192

File: tests/data/reference_graphs/quantized/symmetric/unet_qoutput.dot [hunk 26]
```diff
@@ -1,7 +1,7 @@
 strict digraph  {
```
[the remark this answers is not in the document]
Comment (asenina) [reply tests/data/reference_graphs/quantized/symmetric/unet_qoutput.dot:1]: Done.

File: tests/data/reference_graphs/quantized/symmetric/resnet18_no_qinput.dot [hunk 27]
```diff
@@ -1,247 +1,249 @@
 strict digraph  {
```
[the remark this answers is not in the document]
Comment (asenina) [reply tests/data/reference_graphs/quantized/symmetric/resnet18_no_qinput.dot]: Done.

File: tests/data/reference_graphs/quantized/symmetric/sr_small_model.dot [hunk 28]
```diff
@@ -1,151 +1,287 @@
 strict digraph  {
```
[the remark this answers is not in the document]
Comment (asenina) [reply tests/data/reference_graphs/quantized/symmetric/sr_small_model.dot]: As discussed, add `pytest.skip()` and TODO in code.
Done.

File: tests/quantization/test_algo_quantization.py (hunk 6 above)
Comment (asenina) [reply tests/quantization/test_algo_quantization.py:451]: Done.

File: tests/helpers.py (hunk 3 above)
Comment (asenina) [reply tests/helpers.py]: Done.

File: tests/quantization/test_algo_quantization.py (hunk 4 above)
Comment (asenina) [reply tests/quantization/test_algo_quantization.py:212]: Done.

File: tests/quantization/test_precision_init.py (hunk 10 above)
Comment (asenina) [reply tests/quantization/test_precision_init.py:451]: As discussed, this test originaly had a bug.  It shouldn't work with VPU config.
Done.

File: nncf/quantization/init_precision.py (hunk 2 above)
Comment (ljaljushkin) [reply nncf/quantization/init_precision.py]: why warnings? just to not fail tests? I'd prefer disabling/skipping tests then. The right behavior is an error, as we discussed with the team. If a user wants warnings then he should specify `treat_warning_as_errors`=False, which is True, by default.

File: tests/test_sanity_sample.py [hunk 29]
```diff
@@ -389,6 +389,7 @@ def test_get_default_weight_decay(algo, ref_weight_decay):
 
 def test_cpu_only_mode_produces_cpu_only_model(config, tmp_path, mocker):
     config_factory = ConfigFactory(config['nncf_config'], tmp_path / 'config.json')
+    #config['nncf_config']['device'] = 'cpu'
```
Comment (vshampor) [inline tests/test_sanity_sample.py:392]: No commented code allowed

File: tests/test_compressed_graph.py [hunk 30]
```diff
@@ -152,7 +153,8 @@ def forward_fn(model, seq_len_, batch_size_, vocab_size_, batch_first_):
 
         def gen_packed_sequence():
             seq_list = []
-            seq_lens = torch.LongTensor(batch_size_).random_(1, seq_len_ + 1).to(device)
+            #seq_lens = torch.LongTensor(batch_size_).random_(1, seq_len_ + 1).to(device)
```
Comment (vshampor) [inline tests/test_compressed_graph.py:156]: No commented code allowed

**Review by vshampor — CHANGES_REQUESTED** [review]

**Review by vshampor — APPROVED** [review]
